### ISSUE TITLE
Panic when opening Alsa soundcard on Rodio

### ISSUE THREAD

**Issue by lukasdietz** [librespot-org/librespot#760]
Hey there,

I am facing an issue using librespot. 
I have installed librespot using `cargo install librespot` and used it together with snapcast. 
Everything worked great for around 2 weeks before it wouldn't work anymore. 

If I start librespot everything works as long as nothing happens. If I am trying to connect Librespot with Spotify it panics right away:

```
[2021-05-26T13:17:17Z INFO  librespot] librespot 0.2.0 UNKNOWN (Built on 2021-05-26, Build ID: FooKApaN)
[2021-05-26T13:17:25Z INFO  librespot_core::session] Connecting to AP "gew1-accesspoint-b-d9lk.ap.spotify.com:4070"
[2021-05-26T13:17:25Z INFO  librespot_core::session] Authenticated as "8sybhylbg9kotiukantktue31" !
[2021-05-26T13:17:25Z INFO  librespot_core::session] Country: "DE"
[2021-05-26T13:17:25Z INFO  librespot_playback::audio_backend::rodio] Using Rodio sink with format S16 and cpal host: ALSA
[2021-05-26T13:17:25Z INFO  librespot_playback::audio_backend::rodio] Using audio device: default
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5047:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2565:(snd_pcm_open_noupdate) Unknown PCM default
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: StreamError(NoDevice)', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/librespot-playback-0.2.0/src/audio_backend/rodio.rs:165:53
stack backtrace:
   0:     0x55f3c8416290 - std::backtrace_rs::backtrace::libunwind::trace::h5e9d00f0cdf4f57e
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x55f3c8416290 - std::backtrace_rs::backtrace::trace_unsynchronized::hd5302bd66215dab9
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55f3c8416290 - std::sys_common::backtrace::_print_fmt::ha0237cd11a34e2bf
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x55f3c8416290 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h171d4c10df1a98ee
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x55f3c843b0ec - core::fmt::write::h89e4288724daa3fa
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/fmt/mod.rs:1096:17
   5:     0x55f3c840ea52 - std::io::Write::write_fmt::h6d40f996e84584d9
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/io/mod.rs:1568:15
   6:     0x55f3c8418725 - std::sys_common::backtrace::_print::h0c0b93221682afc8
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x55f3c8418725 - std::sys_common::backtrace::print::h57a9f95204c2fdd6
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x55f3c8418725 - std::panicking::default_hook::{{closure}}::h4245258b50e37e69
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:208:50
   9:     0x55f3c8418283 - std::panicking::default_hook::h7b00dcc1d0944747
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:225:9
  10:     0x55f3c8418ec1 - std::panicking::rust_panic_with_hook::h71e6a073d87de1f5
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:591:17
  11:     0x55f3c8418a07 - std::panicking::begin_panic_handler::{{closure}}::hd549436f6bb6dbb8
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:497:13
  12:     0x55f3c841672c - std::sys_common::backtrace::__rust_end_short_backtrace::h4e5f4b72b04174c3
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:141:18
  13:     0x55f3c8418969 - rust_begin_unwind
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:493:5
  14:     0x55f3c8438861 - core::panicking::panic_fmt::hcd56f7f635f62c74
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/panicking.rs:92:14
  15:     0x55f3c8438683 - core::option::expect_none_failed::hf37eebedadde29e6
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/option.rs:1300:5
  16:     0x55f3c8130058 - librespot_playback::audio_backend::rodio::mk_rodio::h171789833993c2dc
  17:     0x55f3c801827b - std::sys_common::backtrace::__rust_begin_short_backtrace::hdcf24dc2879da306
  18:     0x55f3c8051dec - core::ops::function::FnOnce::call_once{{vtable.shim}}::h308505981f8cf5ba
  19:     0x55f3c842064a - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h61144a2be4ee36d8
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/alloc/src/boxed.rs:1521:9
  20:     0x55f3c842064a - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hcf5d395fdd120c17
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/alloc/src/boxed.rs:1521:9
  21:     0x55f3c842064a - std::sys::unix::thread::Thread::new::thread_start::hb5e40d3d934ebb7a
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys/unix/thread.rs:71:17
  22:     0x7f2bfa682fa3 - start_thread
  23:     0x7f2bfa4294cf - clone
  24:                0x0 - <unknown>
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: SendError(Load(SpotifyId { id: <SpotifyID>, audio_type: Track }, false, 1228))', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/librespot-playback-0.2.0/src/player.rs:334:51
stack backtrace:
   0:     0x55f3c8416290 - std::backtrace_rs::backtrace::libunwind::trace::h5e9d00f0cdf4f57e
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x55f3c8416290 - std::backtrace_rs::backtrace::trace_unsynchronized::hd5302bd66215dab9
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55f3c8416290 - std::sys_common::backtrace::_print_fmt::ha0237cd11a34e2bf
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x55f3c8416290 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h171d4c10df1a98ee
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x55f3c843b0ec - core::fmt::write::h89e4288724daa3fa
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/fmt/mod.rs:1096:17
   5:     0x55f3c840ea52 - std::io::Write::write_fmt::h6d40f996e84584d9
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/io/mod.rs:1568:15
   6:     0x55f3c8418725 - std::sys_common::backtrace::_print::h0c0b93221682afc8
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x55f3c8418725 - std::sys_common::backtrace::print::h57a9f95204c2fdd6
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x55f3c8418725 - std::panicking::default_hook::{{closure}}::h4245258b50e37e69
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:208:50
   9:     0x55f3c8418283 - std::panicking::default_hook::h7b00dcc1d0944747
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:225:9
  10:     0x55f3c8418ec1 - std::panicking::rust_panic_with_hook::h71e6a073d87de1f5
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:591:17
  11:     0x55f3c8418a07 - std::panicking::begin_panic_handler::{{closure}}::hd549436f6bb6dbb8
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:497:13
  12:     0x55f3c841672c - std::sys_common::backtrace::__rust_end_short_backtrace::h4e5f4b72b04174c3
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:141:18
  13:     0x55f3c8418969 - rust_begin_unwind
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:493:5
  14:     0x55f3c8438861 - core::panicking::panic_fmt::hcd56f7f635f62c74
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/panicking.rs:92:14
  15:     0x55f3c8438683 - core::option::expect_none_failed::hf37eebedadde29e6
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/option.rs:1300:5
  16:     0x55f3c810dd8f - librespot_playback::player::Player::load::ha0c442d0f993bc55
  17:     0x55f3c80b4c5a - librespot_connect::spirc::SpircTask::load_track::h0e341e1961cf6e70
  18:     0x55f3c80b23eb - librespot_connect::spirc::SpircTask::handle_frame::h5d4baec7b99df48b
  19:     0x55f3c8039d12 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h91f4475b0d1a5aac
  20:     0x55f3c80370a3 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h45d0cb9326ea4322
  21:     0x55f3c801999b - <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll::h99bfec52701b0988
  22:     0x55f3c8044be7 - librespot::main::{{closure}}::hf174eb5dadaa6a1e
  23:     0x55f3c8002f67 - tokio::macros::scoped_tls::ScopedKey<T>::set::h5afe8179ff76e611
  24:     0x55f3c7fe644f - tokio::runtime::basic_scheduler::BasicScheduler<P>::block_on::h0468d0126abaf00b
  25:     0x55f3c8003aeb - tokio::runtime::Runtime::block_on::h21c5ac6de7a02619
  26:     0x55f3c7fee7a0 - librespot::main::hfa2dab2558bf6cc3
  27:     0x55f3c8018673 - std::sys_common::backtrace::__rust_begin_short_backtrace::heea75a7113230c2d
  28:     0x55f3c807f6b9 - std::rt::lang_start::{{closure}}::h0bed97e3334e6172
  29:     0x55f3c84193d7 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h527fb2333ede305e
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ops/function.rs:259:13
  30:     0x55f3c84193d7 - std::panicking::try::do_call::h309d8aee8149866c
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:379:40
  31:     0x55f3c84193d7 - std::panicking::try::h75a60c31fd16bfc6
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:343:19
  32:     0x55f3c84193d7 - std::panic::catch_unwind::h1f9892423e99bc00
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panic.rs:431:14
  33:     0x55f3c84193d7 - std::rt::lang_start_internal::hd5b67df56ca01dae
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/rt.rs:51:25
  34:     0x55f3c7fee862 - main
  35:     0x7f2bfa35409b - __libc_start_main
  36:     0x55f3c7fd60ca - _start
  37:                0x0 - <unknown>
[2021-05-26T13:17:26Z ERROR librespot_playback::player] Player thread panicked!
```

Strange thing for me is, that it stopped working without me changing anything.
Hope there is somebody to help me out! 

With best regards
Lukas

**Comment by roderickvd** [comment]
The error means that it cannot find or open the soundcard on Alsa.

When you say "without me changing anything" you confirm:
- You were running `librespot` 0.2.0 before? The log says you built it on 2021-05-26 so this cannot be the same as 2 weeks before. 
- You did not perform any system updates?
- You did not change any configuration files (including `/etc/group`)?

Besides the questions above, please do the following:
- try running with `librespot --backend alsa` instead of Rodio (compile with `cargo install librespot --features alsa-backend`)
- report the output of `aplay -L`

**Comment by lukasdietz** [comment]
Sorry for not beeing clear. I forgot to mention, that I reinstalled Librespot today as a first try to fix this on my own.
But for all your other questions: Yes I did run librespot 0.2.0 before and did not change any configuration or run updates.
The only thing I could mention was a reboot of my proxmox host where the lxc with librespot is running on.

I tried compiling librespot as mentioned. Now I can start librespot and connect with spotify without crashing.

The output of `aplay -L` is 
```
null
Discard all samples (playback) or generate zero samples (capture)
```
